File: src/main/kotlin/dev/slne/surf/essentialsold/commands/minecraft/SpectateCommand.java
package dev.slne.surf.essentialsold.commands.minecraft;

import dev.jorel.commandapi.exceptions.WrapperCommandSyntaxException;
import dev.jorel.commandapi.executors.NativeResultingCommandExecutor;
import dev.slne.surf.essentialsold.commands.EssentialsCommand;
import dev.slne.surf.essentialsold.utils.EssentialsUtil;
import dev.slne.surf.essentialsold.utils.color.Colors;
import dev.slne.surf.essentialsold.utils.brigadier.Exceptions;
import dev.slne.surf.essentialsold.utils.permission.Permissions;
import lombok.val;
import net.kyori.adventure.text.Component;
import org.bukkit.GameMode;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;

import javax.annotation.Nullable;

public class SpectateCommand extends EssentialsCommand {
    public SpectateCommand() {
        super("spectate", "spectate [<target>] [<player>]", "Spectates the given player");

        withRequirement(EssentialsUtil.checkPermissions(Permissions.SPECTATE_SELF_PERMISSION, Permissions.SPECTATE_SELF_PERMISSION, Permissions.SPECTATE_OTHER_PERMISSION));

        executesNative((NativeResultingCommandExecutor) (sender, args) -> spectate(
                sender.getCallee(),
                null,
                getPlayerOrException(sender)
        ));
        then(entityArgument("target")
                .withRequirement(EssentialsUtil.checkPermissions(Permissions.SPECTATE_SELF_PERMISSION, Permissions.SPECTATE_OTHER_PERMISSION))
                .executesNative((NativeResultingCommandExecutor) (sender, args) -> spectate(
                        sender.getCallee(),
                        args.getUnchecked("target"),
                        getPlayerOrException(sender)
                ))
                .then(playerArgument("player")
                        .withPermission(Permissions.SPECTATE_OTHER_PERMISSION)
                        .executesNative((NativeResultingCommandExecutor) (sender, args) -> spectate(
                                sender.getCallee(),
                                args.getUnchecked("target"),
                                args.getUnchecked("player")
                        ))
                )
        );
    }

    private int spectate(CommandSender source, @Nullable Entity entityUnchecked, Player playerUnchecked) throws WrapperCommandSyntaxException {
        val entity = EssentialsUtil.checkEntitySuggestion(source, entityUnchecked);
        val player = EssentialsUtil.checkPlayerSuggestion(source, playerUnchecked);

        if (entity != null && entity.equals(player)) throw Exceptions.ERROR_CANNOT_SPECTATE_SELF;

        if (player.getGameMode() != GameMode.SPECTATOR && entity != null) {
            player.setGameMode(GameMode.SPECTATOR);
        }
        player.setSpectatorTarget(entity);

        if (entity != null) {
            EssentialsUtil.sendSuccess(source, EssentialsUtil.getDisplayName(player)
                    .append(Component.text(" beobachtet nun ", Colors.SUCCESS))
                    .append(EssentialsUtil.getDisplayName(entity))
                    .append(Component.text(".", Colors.SUCCESS)));
        } else {
            EssentialsUtil.sendSuccess(source, EssentialsUtil.getDisplayName(player)
                    .append(Component.text(" beobachtet nun niemanden mehr.", Colors.SUCCESS)));
        }
        return 1;
    }
}

File: src/main/kotlin/dev/slne/surf/essentials/command/minecraft/SpectateCommand.kt
package dev.slne.surf.essentials.command.minecraft

import dev.jorel.commandapi.kotlindsl.commandTree
import dev.jorel.commandapi.kotlindsl.entitySelectorArgumentOnePlayer
import dev.jorel.commandapi.kotlindsl.getValue
import dev.jorel.commandapi.kotlindsl.playerExecutor
import dev.slne.surf.essentials.util.permission.EssentialsPermissionRegistry
import dev.slne.surf.surfapi.core.api.messages.adventure.sendText
import org.bukkit.GameMode
import org.bukkit.entity.Player

fun spectateCommand() = commandTree("spectate") {
    withPermission(EssentialsPermissionRegistry.SPECTATE_COMMAND)
    playerExecutor { player, _ ->
        val target = player.spectatorTarget ?: run {
            player.sendText {
                appendPrefix()
                error("Du beobachtest gerade keinen Spieler.")
            }
            return@playerExecutor
        }

        player.spectatorTarget = null

        player.sendText {
            appendPrefix()
            success("Du beobachtest nun nicht mehr ")
            variableValue(target.name)
            success(".")
        }
    }

    entitySelectorArgumentOnePlayer("user") {
        playerExecutor { player, args ->
            val user: Player by args

            player.gameMode = GameMode.SPECTATOR
            player.spectatorTarget = user

            player.sendText {
                appendPrefix()
                success("Du beobachtest nun ")
                variableValue(user.name)
                success(".")
            }
        }

        entitySelectorArgumentOnePlayer("target") {
            withPermission(EssentialsPermissionRegistry.SPECTATE_COMMAND_OTHERS)
            playerExecutor { executor, args ->
                val user: Player by args
                val target: Player by args

                user.gameMode = GameMode.SPECTATOR
                user.spectatorTarget = target

                executor.sendText {
                    appendPrefix()
                    variableValue(user.name)
                    success(" beobachtet nun ")
                    variableValue(target.name)
                    success(".")
                }

                user.sendText {
                    appendPrefix()
                    success("Du beobachtest nun ")
                    variableValue(target.name)
                    success(".")
                }
            }
        }
    }
}